# HW1

## Problem 1

In [0]:
# Code from Chapter 2 of Machine Learning: An Algorithmic Perspective
# by Stephen Marsland (http://seat.massey.ac.nz/personal/s.r.marsland/MLBook.html)

# You are free to use, change, or redistribute the code in any way you wish for
# non-commercial purposes, but please maintain the name of the original author.
# This code comes with no warranty of any kind.

# Stephen Marsland, 2008

from numpy import *

class pcn:
	""" A basic Perceptron (the same pcn.py except with the weights printed
	and it does not reorder the inputs)"""
	
	def __init__(self,inputs,targets):
		""" Constructor """
		# Set up network size
		if ndim(inputs)>1:
			self.nIn = shape(inputs)[1]
		else: 
			self.nIn = 1
	
		if ndim(targets)>1:
			self.nOut = shape(targets)[1]
		else:
			self.nOut = 1

		self.nData = shape(inputs)[0]
	
		# Initialise network
		self.weights = random.rand(self.nIn+1,self.nOut)*0.1-0.05

	def pcntrain(self,inputs,targets,eta,nIterations):
		""" Train the thing """	
		# Add the inputs that match the bias node
		inputs = concatenate((inputs,-ones((self.nData,1))),axis=1)
	
		# Training
		change = range(self.nData)

		for n in range(nIterations):
			
			self.outputs = self.pcnfwd(inputs);
			self.weights += eta*dot(transpose(inputs),targets-self.outputs)
			print ("Iteration: ", n)
			print (self.weights)
			
			activations = self.pcnfwd(inputs)
			print ("Final outputs are:")
			print (activations)
		#return self.weights

	def pcnfwd(self,inputs):
		""" Run the network forward """

		outputs =  dot(inputs,self.weights)

		# Threshold the outputs
		return where(outputs>0.5,1,0)


	def confmat(self,inputs,targets):
		"""Confusion matrix"""

		# Add the inputs that match the bias node
		inputs = concatenate((inputs,-ones((self.nData,1))),axis=1)
		outputs = dot(inputs,self.weights)
	
		nClasses = shape(targets)[1]

		if nClasses==1:
			nClasses = 2
			outputs = where(outputs>0,1,0)
		else:
			# 1-of-N encoding
			outputs = argmax(outputs,1)
			targets = argmax(targets,1)

		cm = zeros((nClasses,nClasses))
		for i in range(nClasses):
			for j in range(nClasses):
				cm[i,j] = sum(where(outputs==i,1,0)*where(targets==j,1,0))

		print (cm)
		print (trace(cm)/sum(cm))

**Test Case**

In [9]:
import numpy as np
inputs = np.array([[0,0],[0,1],[1,0],[1,1]])
targets = np.array([[0],[1],[1],[1]])

p = pcn(inputs, targets)
p.pcntrain(inputs, targets, 0.1, 6)

Iteration:  0
[[ 0.24458579]
 [ 0.20418874]
 [-0.2809949 ]]
Final outputs are:
[[0]
 [0]
 [1]
 [1]]
Iteration:  1
[[ 0.24458579]
 [ 0.30418874]
 [-0.3809949 ]]
Final outputs are:
[[0]
 [1]
 [1]
 [1]]
Iteration:  2
[[ 0.24458579]
 [ 0.30418874]
 [-0.3809949 ]]
Final outputs are:
[[0]
 [1]
 [1]
 [1]]
Iteration:  3
[[ 0.24458579]
 [ 0.30418874]
 [-0.3809949 ]]
Final outputs are:
[[0]
 [1]
 [1]
 [1]]
Iteration:  4
[[ 0.24458579]
 [ 0.30418874]
 [-0.3809949 ]]
Final outputs are:
[[0]
 [1]
 [1]
 [1]]
Iteration:  5
[[ 0.24458579]
 [ 0.30418874]
 [-0.3809949 ]]
Final outputs are:
[[0]
 [1]
 [1]
 [1]]


## Problem 2


In [2]:
import numpy as np

def linreg(inputs,targets):
  inputs = np.concatenate((inputs,-np.ones((np.shape(inputs)[0],1))), axis=1)
  beta = np.dot(np.dot(np.linalg.inv(np.dot(np.transpose(inputs), inputs)),np.transpose(inputs)),targets)
  
  outputs = np.dot(inputs,beta)
  
  return beta

trainin = [[28540],
           [40133], 
           [39900], 
           [0],
           [0], 
           [42050],
           [43220], 
           [39565]]
traintgt = [137, 135, 127, 122, 120, 118, 118, 117]
testin = [[40400], 
          [54506]]
testtgt = [117, 114]

beta = linreg(trainin,traintgt)
testin = np.concatenate((testin,-np.ones((np.shape(testin)[0],1))), axis=1) 
testout = np.dot(testin,beta)
error = np.sum((testout - testtgt)**2)

print(testout)
print(error)

[124.64724943 125.14650108]
182.7249102032826


According to the running result above, we can find we actully have a output aroung $125$ which means the answer does not useful. According to the data point we have now, we can say that higher cost of tuition translate does **NOT** have higher-paying jobs 

## Problem 3
Because our result is binary number then we get

$logit = \log{\frac{P(Y=1|X)}{1-P(Y=1|X)}}=X\beta$

And, $X\beta=w_1x_1+w_2x_2+\cdots+w_nx_n$

Thus, $P(Y=1|X)=\frac{1}{1+e^{-X\beta}}$ and $X=[AGE, OBS]$

## Problem 4
**a)** To use an MLP in this case, I consider season and tempratrue are very important and also weekday and weekends affects a lot. I think season is the most important one and can have the largetest wight. Then comes to weekday or weekends. The least important I believe is temprature.

**b)** If this two facts added to system, I'll consider add them as a prority factor in system, because most electronic devices are switch off or in low power mode during night time. Also temprature of that perior will decide people to open AC or hearter or not, which both are going to use a lot of energy.

**c)** I believe this MLP will work well. The reason is that electronic usage is always high in winter and summer and low in other two. High in day time and low in night. There are a lot of relulation to obey. Thus we can actully train a usful model.

## Problem 5
I will choose 4 inputs and 3 hidden layer. For inputs, they arenumber of people in the geriatric ward each week, weather, season, and epidemic status. For hidden layer, I'll create three for first three inputs. Beasue this value may affect each others. Comes to epidemic status, I believe it will derictly affect result, then I'll connect them directly. I think this system will work, because we've consider most of conditions which are affecting the result. If we choose sensible parameters, we can produce a even prefect result.

## Problem 6
**1.** The new point **U** should be in color **RED**.

**2.** The answer is one of **Yellow**, **Purple**, or **Green**. After we change the distence function, we acctully find the farthest points. And all three points above are the three farthest points, and the make it a tie. Thus, the specific answer will based on our algorithm which deal with tie condition.